lib/ron.nix

```diff
@@ -0,0 +1,284 @@
+              _SKIP_ rdelim]);
+
+    _LIST_ = _OBJECTS_ _LBRACKET_ _RBRACKET_;
+    _TUPLE_ = bind
+      (v: let
+        type = head v;
+        elements = elemAt v 1;
+      in
+        if type == null then RON.tuple elements
+        else RON.tupleWithType (head type) elements)
+      (AND [(OPT (AND [_NAME_ _SKIP_]))
+            (_OBJECTS_ _LPAREN_ _RPAREN_)]);
+
+    _FIELD_ = bind (v: { name = (head v); value = (elemAt v 4); })
+      (AND [_NAME_ _SKIP_ _COLON_ _SKIP_ _OBJECT_]);
+    _STRUCT_ = bind
+      (v: let
+        type = head v;
+        fs = elemAt v 3;
+        value = listToAttrs (if fs == null then []
+                             else [(head fs)] ++ (assembleList (v: elemAt v 3) (elemAt fs 1)));
+      in
+        if type == null then RON.struct value
+        else RON.structWithType (head type) value)
+      (AND [(OPT (AND [_NAME_ _SKIP_]))
+            _LPAREN_ _SKIP_
+            (OPT (AND [_FIELD_
+                       (STAR (AND [_SKIP_ _COMMA_ _SKIP_ _FIELD_]))]))
+            (OPT (AND [_SKIP_ _COMMA_]))
+            _SKIP_ _RPAREN_]);
+
+    _MAPPING_ = bind (v: { key = (head v); value = (elemAt v 4); })
+      (AND [_OBJECT_ _SKIP_ _COLON_ _SKIP_ _OBJECT_]);
+    _MAP_ = bind
+      (v: let
+        ms = elemAt v 2;
+      in
+        RON.map (if ms == null then []
+                 else [(head ms)] ++ (assembleList (v: elemAt v 3) (elemAt ms 1))))
+      (AND [_LCURLY_ _SKIP_
+            (OPT (AND [_MAPPING_
+                       (STAR (AND [_SKIP_ _COMMA_ _SKIP_ _MAPPING_]))]))
+            (OPT (AND [_SKIP_ _COMMA_]))
+            _SKIP_ _RCURLY_]);
+
+    _OBJECT_ = OR [_MAP_ _UNIT_ _SOME_ _TUPLE_ _LIST_ _STRUCT_ _NONE_ _IDENTIFIER_ _STRING_ _NUM_ _BOOL_];
+
+    r = AND [_SKIP_ _OBJECT_ _SKIP_] 0;
+  in
+    if r == null then throw "RON parse error"
+    else if r.end < len then throw "Junk after position ${toString r.end}"
+    else elemAt r.value 1;
+
+  toRON = o:
+    if o == null then "()"
+    else if isBool o || isInt o then "${toString o}"
+    else if isString o then "\"${o}\""
+    else if isList o then "[${concatStringsSep ", " (map toRON o)}]"
+    else if isAttrs o then
+      if RON.typeOf o == "identifier" then "${o.name}"
+      else if RON.typeOf o == "some" then "Some(${toRON o.value})"
+      else if RON.typeOf o == "none" then "None"
+      else if RON.typeOf o == "tuple" then "${o.__RONTupleType or ""}(${concatStringsSep ", " (map toRON o.elements)})"
+      else if RON.typeOf o == "struct" then "${o.__RONStructType or ""}(${concatStringsSep ", "
+        (map (key: "${key}: ${toRON o.${key}}") (filter (n: n != "__RONStructType") (attrNames o)))})"
+      else if RON.typeOf o == "map" then "{${concatStringsSep ", " (map (m: "${toRON m.key}: ${toRON m.value}") o.mappings)}}"
+      else throw "Don't know how to serialize \"${RON.typeOf o}\" to RON"
+    else throw "Don't know how to serialize \"${typeOf o}\" to RON";
+in {
+  inherit RON fromRON toRON;
```

Comment:
I don't think adding this kind of code to nixpkgs is a good idea. While it's neat that it works at all, Nix isn't meant for writing parsers and serializers. The code here is near impossible to maintain and comes with no tests. Furthermore it seems awful to use, since you need to know about all the RON specific Nix functions in addition to RON itself to create a valid configuration file.

Reply:
Are you aware of, say, this: https://github.com/talyz/fromElisp/blob/master/default.nix ?
Nix is actually surprisingly handy at parsing -- it being a lazy functional language is enough of a boon and its shortcomings in string processing and matching are not enough of an impediment.

I fear this RON abomination is going to be very common soon, what with all the Rust hype, and in a few years' time `builtins.fromRON` is going to be a thing.

That all said, your points about maintenability and tests is well-taken. I'll redo this PR in a less ambitious way.

Reply:
> and in a few years' time builtins.fromRON is going to be a thing.

It needs a sanctioned grammar first, and well, Rust is far from being a stable language...